### PR TITLE
update(C04_P09): used deque to avoid slicing

### PR DESCRIPTION
**Pull request by gan3i**
deque gives the ability to remove and add the element at the  beginning of the sequences in constant time since it is implemented using linked list, this also avoids slicing of the sequences on every recursive call.

### CODE REVIEW

File: chapter_04/p09_bst_sequences.py [hunk 1]
```diff
@@ -1,5 +1,7 @@
 from chapter_04.binary_search_tree import BinarySearchTree
 
+from collections import deque
+
```
Comment (brycedrennan) [inline chapter_04/p09_bst_sequences.py:4]: > Imports are always put at the top of the file, just after any module comments and docstrings, and before module globals and constants.
Imports should be grouped in the following order:
> - Standard library imports.
> - Related third party imports.
> - Local application/library specific imports.
>
> You should put a blank line between each group of imports.


https://www.python.org/dev/peps/pep-0008/#imports

Comment (brycedrennan) [reply chapter_04/p09_bst_sequences.py:4]:
```python
from collections import deque

from chapter_04.binary_search_tree import BinarySearchTree
```

Comment (gan3i) [reply chapter_04/p09_bst_sequences.py:2]: Thanks, I have made the corresponding changes.

**Comment by gan3i** [comment]
Hi @brycedrennan any thoughts on what is the time and space complexity of this algorithm? it seems exponential to me.

**Comment by brycedrennan** [comment]
@gan3i thanks for this improvement.   

Your changes here make sense and I would have thought they make things faster but in testing I was surprised to see this changes appears to make things slower. 

image

I was able to recover most, but not all, of the performance by making one adjustment to your PR in the helper function:

```python
            d = deque()
            d.append(node.key)
            sequences = weave(left, right, d, sequences)
```
instead of
```python
            sequences = weave(left, right, deque([node.key]), sequences)
```

This avoids instantiating the list `[node.key]`


This is the code I used:
```python
def test_perf():
    for n in [2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]:
        start = time.perf_counter()
        for run in range(100):
            bst = BinarySearchTree()
            for _ in range(n):
                bst.insert(random.randint(1, 10000))
            sequences = find_bst_sequences(bst)
        duration = (time.perf_counter() - start) * 1000
        print(f"{n}, {duration}")
```

Since this doesn't improve the performance I don't think we should merge it as is.  If you find a way to make this faster I'm happy to reopen this pull request.

**Comment by brycedrennan** [comment]
And yes it does look like it's exponential.

**Pull request closed without merging**